File: erpnext/construction_inspection/report/inspection_summary/inspection_summary.js
// Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Inspection Summary"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "construction_site",
			"label": __("Construction Site"),
			"fieldtype": "Link",
			"options": "Construction Site"
		},
		{
			"fieldname": "supplier",
			"label": __("Supplier"),
			"fieldtype": "Link",
			"options": "Supplier"
		},
		{
			"fieldname": "inspector",
			"label": __("Inspector"),
			"fieldtype": "Data"
		},
		{
			"fieldname": "min_rating",
			"label": __("Minimum Rating"),
			"fieldtype": "Float",
			"precision": 1
		},
		{
			"fieldname": "max_rating",
			"label": __("Maximum Rating"),
			"fieldtype": "Float",
			"precision": 1
		},
		{
			"fieldname": "status",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": "\nDraft\nSubmitted\nApproved\nRejected"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// 評分欄位的顏色格式化
		if (["safety_rating", "quality_rating", "progress_rating", "overall_rating"].includes(column.fieldname)) {
			const rating = parseFloat(data[column.fieldname]);
			if (rating >= 4.5) {
				value = `<span style="color: #28a745; font-weight: bold;">${value}</span>`;
			} else if (rating >= 3.5) {
				value = `<span style="color: #17a2b8; font-weight: bold;">${value}</span>`;
			} else if (rating >= 2.5) {
				value = `<span style="color: #ffc107; font-weight: bold;">${value}</span>`;
			} else if (rating > 0) {
				value = `<span style="color: #dc3545; font-weight: bold;">${value}</span>`;
			}
		}
		
		// 狀態欄位的格式化
		if (column.fieldname === "status") {
			const status = data.status;
			let color = "#6c757d";
			if (status === "Approved") color = "#28a745";
			else if (status === "Submitted") color = "#17a2b8";
			else if (status === "Rejected") color = "#dc3545";
			
			value = `<span style="color: ${color}; font-weight: bold;">${value}</span>`;
		}
		
		// 問題數量的格式化
		if (column.fieldname === "issues_count") {
			const count = parseInt(data.issues_count);
			if (count > 0) {
				value = `<span style="color: #dc3545; font-weight: bold;">${count}</span>`;
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// 添加自定義按鈕
		report.page.add_inner_button(__("Export to Excel"), function() {
			frappe.query_report.export_report();
		});
		
		report.page.add_inner_button(__("Analytics Dashboard"), function() {
			frappe.set_route("query-report", "Inspection Analytics");
		});
		
		report.page.add_inner_button(__("Create Inspection"), function() {
			frappe.new_doc("Supplier Inspection");
		});
		
		// 添加刷新按鈕
		report.page.add_action_icon("fa fa-refresh", function() {
			report.refresh();
		});
	},
	
	"after_datatable_render": function(datatable_obj) {
		// 添加行點擊事件
		$(datatable_obj.wrapper).find('.dt-row').on('click', function() {
			const data = datatable_obj.datamanager.getRow($(this).index());
			if (data && data[0].content) {
				frappe.set_route("Form", "Supplier Inspection", data[0].content);
			}
		});
		
		// 添加工具提示
		$(datatable_obj.wrapper).find('[data-original-title]').tooltip();
	}
};

// 自定義函數：獲取巡檢分析數據
function get_inspection_analytics() {
	const filters = frappe.query_report.get_filter_values();
	
	frappe.call({
		method: "erpnext.construction_inspection.report.inspection_summary.inspection_summary.get_inspection_analytics",
		args: {
			filters: filters
		},
		callback: function(r) {
			if (r.message) {
				show_analytics_dialog(r.message);
			}
		}
	});
}

// 顯示分析對話框
function show_analytics_dialog(data) {
	const dialog = new frappe.ui.Dialog({
		title: __("Inspection Analytics"),
		size: "large",
		fields: [
			{
				fieldtype: "HTML",
				fieldname: "analytics_html"
			}
		]
	});
	
	const stats = data.stats;
	const monthly_trend = data.monthly_trend;
	const supplier_ranking = data.supplier_ranking;
	
	const html = `
		<div class="analytics-container">
			<div class="row">
				<div class="col-md-6">
					<div class="card">
						<div class="card-header">
							<h5>基本統計</h5>
						</div>
						<div class="card-body">
							<table class="table table-borderless">
								<tr>
									<td>總巡檢次數:</td>
									<td><strong>${stats.total_inspections}</strong></td>
								</tr>
								<tr>
									<td>平均評分:</td>
									<td><strong>${(stats.avg_rating || 0).toFixed(1)}</strong></td>
								</tr>
								<tr>
									<td>優秀評分數:</td>
									<td><strong>${stats.excellent_count}</strong></td>
								</tr>
								<tr>
									<td>待改進數:</td>
									<td><strong>${stats.poor_count}</strong></td>
								</tr>
								<tr>
									<td>涉及現場數:</td>
									<td><strong>${stats.sites_inspected}</strong></td>
								</tr>
								<tr>
									<td>涉及供應商數:</td>
									<td><strong>${stats.suppliers_involved}</strong></td>
								</tr>
							</table>
						</div>
					</div>
				</div>
				<div class="col-md-6">
					<div class="card">
						<div class="card-header">
							<h5>供應商排名 (前10名)</h5>
						</div>
						<div class="card-body">
							<table class="table table-sm">
								<thead>
									<tr>
										<th>供應商</th>
										<th>平均評分</th>
										<th>巡檢次數</th>
									</tr>
								</thead>
								<tbody>
									${supplier_ranking.map((supplier, index) => `
										<tr>
											<td>${supplier.supplier}</td>
											<td><span class="badge badge-${supplier.avg_rating >= 4.5 ? 'success' : supplier.avg_rating >= 3.5 ? 'info' : 'warning'}">${supplier.avg_rating.toFixed(1)}</span></td>
											<td>${supplier.inspection_count}</td>
										</tr>
									`).join('')}
								</tbody>
							</table>
						</div>
					</div>
				</div>
			</div>
		</div>
	`;
	
	dialog.fields_dict.analytics_html.$wrapper.html(html);
	dialog.show();
}

// 添加CSS樣式
frappe.provide("frappe.query_reports");
$(`
<style>
	.analytics-container .card {
		margin-bottom: 20px;
		border: 1px solid #dee2e6;
		border-radius: 8px;
	}
	
	.analytics-container .card-header {
		background-color: #f8f9fa;
		border-bottom: 1px solid #dee2e6;
		padding: 12px 20px;
	}
	
	.analytics-container .card-header h5 {
		margin: 0;
		font-size: 16px;
		font-weight: 600;
	}
	
	.analytics-container .card-body {
		padding: 20px;
	}
	
	.analytics-container .table td {
		padding: 8px 0;
		border: none;
	}
	
	.analytics-container .table th {
		border-bottom: 2px solid #dee2e6;
		font-weight: 600;
	}
	
	.analytics-container .badge {
		font-size: 12px;
		padding: 4px 8px;
	}
</style>
`).appendTo("head");

File: erpnext/construction_inspection/report/inspection_summary/inspection_summary.py
# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, today, add_days


def execute(filters=None):
	"""執行巡檢摘要報表"""
	columns = get_columns()
	data = get_data(filters)
	chart = get_chart_data(data)
	
	return columns, data, None, chart


def get_columns():
	"""定義報表欄位"""
	return [
		{
			"label": _("Construction Site"),
			"fieldname": "construction_site",
			"fieldtype": "Link",
			"options": "Construction Site",
			"width": 200
		},
		{
			"label": _("Supplier"),
			"fieldname": "supplier",
			"fieldtype": "Link",
			"options": "Supplier",
			"width": 150
		},
		{
			"label": _("Inspection Date"),
			"fieldname": "inspection_date",
			"fieldtype": "Date",
			"width": 120
		},
		{
			"label": _("Inspector"),
			"fieldname": "inspector",
			"fieldtype": "Data",
			"width": 120
		},
		{
			"label": _("Safety Rating"),
			"fieldname": "safety_rating",
			"fieldtype": "Float",
			"width": 100,
			"precision": 1
		},
		{
			"label": _("Quality Rating"),
			"fieldname": "quality_rating",
			"fieldtype": "Float",
			"width": 100,
			"precision": 1
		},
		{
			"label": _("Progress Rating"),
			"fieldname": "progress_rating",
			"fieldtype": "Float",
			"width": 100,
			"precision": 1
		},
		{
			"label": _("Overall Rating"),
			"fieldname": "overall_rating",
			"fieldtype": "Float",
			"width": 100,
			"precision": 1
		},
		{
			"label": _("Status"),
			"fieldname": "status",
			"fieldtype": "Data",
			"width": 100
		},
		{
			"label": _("Issues Found"),
			"fieldname": "issues_count",
			"fieldtype": "Int",
			"width": 100
		},
		{
			"label": _("Corrective Actions"),
			"fieldname": "corrective_actions_count",
			"fieldtype": "Int",
			"width": 120
		}
	]


def get_data(filters):
	"""獲取報表數據"""
	conditions = get_conditions(filters)
	
	data = frappe.db.sql(f"""
		SELECT 
			si.construction_site,
			cs.supplier,
			si.inspection_date,
			si.inspector,
			si.safety_compliance_rating as safety_rating,
			si.quality_rating,
			si.progress_rating,
			si.overall_rating,
			si.status,
			(SELECT COUNT(*) FROM `tabInspection Checklist Item` 
			 WHERE parent = si.name AND status = 'Fail') as issues_count,
			(SELECT COUNT(*) FROM `tabInspection Checklist Item` 
			 WHERE parent = si.name AND corrective_action IS NOT NULL 
			 AND corrective_action != '') as corrective_actions_count
		FROM 
			`tabSupplier Inspection` si
		LEFT JOIN 
			`tabConstruction Site` cs ON si.construction_site = cs.name
		WHERE 
			si.docstatus = 1 {conditions}
		ORDER BY 
			si.inspection_date DESC, si.overall_rating DESC
	""", filters, as_dict=1)
	
	return data


def get_conditions(filters):
	"""構建查詢條件"""
	conditions = ""
	
	if filters.get("from_date"):
		conditions += " AND si.inspection_date >= %(from_date)s"
	
	if filters.get("to_date"):
		conditions += " AND si.inspection_date <= %(to_date)s"
	
	if filters.get("construction_site"):
		conditions += " AND si.construction_site = %(construction_site)s"
	
	if filters.get("supplier"):
		conditions += " AND cs.supplier = %(supplier)s"
	
	if filters.get("inspector"):
		conditions += " AND si.inspector = %(inspector)s"
	
	if filters.get("min_rating"):
		conditions += " AND si.overall_rating >= %(min_rating)s"
	
	if filters.get("max_rating"):
		conditions += " AND si.overall_rating <= %(max_rating)s"
	
	if filters.get("status"):
		conditions += " AND si.status = %(status)s"
	
	return conditions


def get_chart_data(data):
	"""生成圖表數據"""
	if not data:
		return None
	
	# 評分分布圖表
	rating_ranges = {
		"優秀 (4.5-5.0)": 0,
		"良好 (3.5-4.4)": 0,
		"一般 (2.5-3.4)": 0,
		"待改進 (1.0-2.4)": 0
	}
	
	for row in data:
		rating = flt(row.get("overall_rating", 0))
		if rating >= 4.5:
			rating_ranges["優秀 (4.5-5.0)"] += 1
		elif rating >= 3.5:
			rating_ranges["良好 (3.5-4.4)"] += 1
		elif rating >= 2.5:
			rating_ranges["一般 (2.5-3.4)"] += 1
		elif rating >= 1.0:
			rating_ranges["待改進 (1.0-2.4)"] += 1
	
	chart = {
		"data": {
			"labels": list(rating_ranges.keys()),
			"datasets": [
				{
					"name": "巡檢次數",
					"values": list(rating_ranges.values())
				}
			]
		},
		"type": "donut",
		"height": 300,
		"colors": ["#28a745", "#17a2b8", "#ffc107", "#dc3545"]
	}
	
	return chart


@frappe.whitelist()
def get_inspection_analytics(filters=None):
	"""獲取巡檢分析數據"""
	if isinstance(filters, str):
		import json
		filters = json.loads(filters)
	
	conditions = get_conditions(filters or {})
	
	# 基本統計
	stats = frappe.db.sql(f"""
		SELECT 
			COUNT(*) as total_inspections,
			AVG(overall_rating) as avg_rating,
			COUNT(CASE WHEN overall_rating >= 4.5 THEN 1 END) as excellent_count,
			COUNT(CASE WHEN overall_rating < 2.5 THEN 1 END) as poor_count,
			COUNT(DISTINCT construction_site) as sites_inspected,
			COUNT(DISTINCT cs.supplier) as suppliers_involved
		FROM 
			`tabSupplier Inspection` si
		LEFT JOIN 
			`tabConstruction Site` cs ON si.construction_site = cs.name
		WHERE 
			si.docstatus = 1 {conditions}
	""", filters or {}, as_dict=1)[0]
	
	# 月度趨勢
	monthly_trend = frappe.db.sql(f"""
		SELECT 
			DATE_FORMAT(inspection_date, '%Y-%m') as month,
			COUNT(*) as inspection_count,
			AVG(overall_rating) as avg_rating
		FROM 
			`tabSupplier Inspection` si
		LEFT JOIN 
			`tabConstruction Site` cs ON si.construction_site = cs.name
		WHERE 
			si.docstatus = 1 
			AND inspection_date >= DATE_SUB(CURDATE(), INTERVAL 12 MONTH)
			{conditions}
		GROUP BY 
			DATE_FORMAT(inspection_date, '%Y-%m')
		ORDER BY 
			month
	""", filters or {}, as_dict=1)
	
	# 供應商排名
	supplier_ranking = frappe.db.sql(f"""
		SELECT 
			cs.supplier,
			COUNT(*) as inspection_count,
			AVG(si.overall_rating) as avg_rating,
			COUNT(CASE WHEN si.overall_rating >= 4.5 THEN 1 END) as excellent_count
		FROM 
			`tabSupplier Inspection` si
		LEFT JOIN 
			`tabConstruction Site` cs ON si.construction_site = cs.name
		WHERE 
			si.docstatus = 1 {conditions}
		GROUP BY 
			cs.supplier
		HAVING 
			COUNT(*) >= 3
		ORDER BY 
			avg_rating DESC, inspection_count DESC
		LIMIT 10
	""", filters or {}, as_dict=1)
	
	return {
		"stats": stats,
		"monthly_trend": monthly_trend,
		"supplier_ranking": supplier_ranking
	}


@frappe.whitelist()
def get_inspection_filters():
	"""獲取報表篩選選項"""
	
	# 獲取施工現場列表
	sites = frappe.get_all("Construction Site", 
		fields=["name", "site_name"], 
		filters={"status": ["!=", "Completed"]},
		order_by="site_name")
	
	# 獲取供應商列表
	suppliers = frappe.get_all("Supplier", 
		fields=["name", "supplier_name"],
		order_by="supplier_name")
	
	# 獲取巡檢員列表
	inspectors = frappe.db.sql("""
		SELECT DISTINCT inspector 
		FROM `tabSupplier Inspection` 
		WHERE inspector IS NOT NULL AND inspector != ''
		ORDER BY inspector
	""", as_dict=1)
	
	return {
		"sites": sites,
		"suppliers": suppliers,
		"inspectors": [{"name": i.inspector} for i in inspectors]
	}